### PR TITLE
Adding Test Case for Issue Multiple Deletes for Shared Backup

### PR DESCRIPTION
**Pull request by skonda-px**
**What this PR does / why we need it**:
Adding Test Case for Issue Multiple Deletes for Shared Backup

**Which issue(s) this PR fixes** (optional)
Closes #PA-531

**Special notes for your reviewer**:

### CODE REVIEW

**Review by snigdha-px — CHANGES_REQUESTED** [review]
Please add aetos link

File: tests/backup/backup_test.go [hunk 1]
```diff
@@ -6200,6 +6200,242 @@ var _ = Describe("{DeleteAllBackupObjects}", func() {
...
+			// Delete the backups
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			var wg sync.WaitGroup
+			backupDriver := Inst().Backup
+			for _, user := range users {
+				wg.Add(1)
+				go func(user string) {
+					defer wg.Done()
+					_, err = DeleteBackup(backupNames[0], backupMap[backupNames[0]], orgID, ctx)
```
Comment (snigdha-px) [inline tests/backup/backup_test.go:6335]: Here context should be user context

File: tests/backup/backup_test.go [hunk 2]
```diff
@@ -6200,6 +6200,242 @@ var _ = Describe("{DeleteAllBackupObjects}", func() {
...
+							return "", false, nil
+						}
+						task.DoRetryWithTimeout(restoreSuccessCheck, 10*time.Minute, 30*time.Second)
+						restoreInspectRequest := &api.RestoreInspectRequest{
+							Name:  restore,
+							OrgId: orgID,
+						}
+						resp, _ := backupDriver.InspectRestore(ctxNonAdmin, restoreInspectRequest)
+						dash.VerifyFatal(resp.GetRestore().GetStatus().Status, api.RestoreInfo_StatusInfo_Success, "Inspecting the Restore success for - "+resp.GetRestore().GetName())
+					}
```
Comment (snigdha-px) [inline tests/backup/backup_test.go:6378]: Above 4 lines are not needed i think

File: tests/backup/backup_test.go [hunk 3]
```diff
@@ -6200,6 +6200,242 @@ var _ = Describe("{DeleteAllBackupObjects}", func() {
...
+		log.FailOnError(err, "Fetching px-central-admin ctx")
+
+		log.Infof("Deleting registered clusters for admin context")
+		DeleteCluster(SourceClusterName, orgID, ctx)
+		DeleteCluster(destinationClusterName, orgID, ctx)
+
+		log.Infof("Deleting registered clusters for non-admin context")
+		for _, ctxNonAdmin := range userContexts {
+			DeleteCluster(SourceClusterName, orgID, ctxNonAdmin)
+			DeleteCluster(destinationClusterName, orgID, ctxNonAdmin)
```
Comment (snigdha-px) [inline tests/backup/backup_test.go:6426]: we can use 'DeleteCloudAccounts' instead of the 2 lines

File: tests/backup/backup_test.go [hunk 4]
```diff
@@ -6200,6 +6200,242 @@ var _ = Describe("{DeleteAllBackupObjects}", func() {
...
+			}(userName)
+		}
+		wg.Wait()
+
+		ctx, err := backup.GetAdminCtxFromSecret()
+		log.FailOnError(err, "Fetching px-central-admin ctx")
+
+		log.Infof("Deleting registered clusters for admin context")
+		DeleteCluster(SourceClusterName, orgID, ctx)
+		DeleteCluster(destinationClusterName, orgID, ctx)
```
Comment (snigdha-px) [inline tests/backup/backup_test.go:6421]: we can use 'DeleteCloudAccounts' instead of the 2 lines

**Review by snigdha-px — CHANGES_REQUESTED** [review]

File: tests/backup/backup_test.go (hunk 2 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:6378]: Discussed with skonda. This is needed

**Comment by skonda-px** [comment]
> Please add aetos link

http://aetos.pwx.purestorage.com/resultSet/testSetID/85785

File: tests/backup/backup_test.go [hunk 5]
```diff
@@ -6200,6 +6200,242 @@ var _ = Describe("{DeleteAllBackupObjects}", func() {
...
+				CloudCredUidList = append(CloudCredUidList, cloudCredUID)
+				backupLocationUID = uuid.New()
+				credName = fmt.Sprintf("autogenerated-cred-%v", time.Now().Unix())
+				CreateCloudCredential(provider, credName, cloudCredUID, orgID)
+				log.InfoD("Created Cloud Credentials with name - %s", credName)
+				//TODO: Eliminate time.Sleep
+				time.Sleep(time.Minute * 1)
+				BackupLocationName = fmt.Sprintf("autogenerated-backup-location-%v", time.Now().Unix())
+				CreateBackupLocation(provider, BackupLocationName, backupLocationUID, credName, cloudCredUID, BucketName, orgID, "")
+				log.InfoD("Created Backup Location with name - %s", BackupLocationName)
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:6264]: are we not returing any error here?

File: tests/backup/backup_test.go [hunk 6]
```diff
@@ -6200,6 +6200,242 @@ var _ = Describe("{DeleteAllBackupObjects}", func() {
...
+			}
+		})
+		Step("Register source and destination cluster for backup", func() {
+			log.InfoD("Registering Source and Destination clusters and verifying the status")
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			CreateSourceAndDestClusters(orgID, "", "", ctx)
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:6271]: same here

File: tests/backup/backup_test.go [hunk 7]
```diff
@@ -6200,6 +6200,242 @@ var _ = Describe("{DeleteAllBackupObjects}", func() {
...
+				for _, restore := range restoreNames {
+					log.Infof("Validting Restore %s for user %s", restore, user)
+					if strings.Contains(restore, user) {
+						restoreSuccessCheck := func() (interface{}, bool, error) {
+							restoreInspectRequest := &api.RestoreInspectRequest{
+								Name:  restore,
+								OrgId: orgID,
+							}
+							resp, err := Inst().Backup.InspectRestore(ctxNonAdmin, restoreInspectRequest)
+							restoreResponseStatus := resp.GetRestore().GetStatus()
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:6360]: handle the err here

File: tests/backup/backup_test.go [hunk 8]
```diff
@@ -6200,6 +6200,242 @@ var _ = Describe("{DeleteAllBackupObjects}", func() {
...
+							log.FailOnError(err, "Failed verifying restore for - %s", restore)
+							actual := restoreResponseStatus.GetStatus()
+							expected := api.RestoreInfo_StatusInfo_Success
+							if actual != expected {
+								log.Infof("Restore status - %s", restoreResponseStatus)
+								log.InfoD("Status of %s - [%s]", restore, restoreResponseStatus.GetStatus())
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:6366]: check if you are logging twice

File: tests/backup/backup_test.go (hunk 7 above)
Comment (lsrinivas-pure) [reply tests/backup/backup_test.go:6360]: return the err as you are in retry logic

File: tests/backup/backup_test.go [hunk 9]
```diff
@@ -6200,6 +6200,242 @@ var _ = Describe("{DeleteAllBackupObjects}", func() {
...
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			for _, restore := range restoreNames {
+				if strings.Contains(restore, user) {
+					log.Infof("deleting Restore %s for user %s", restore, user)
+					DeleteRestore(restore, orgID, ctxNonAdmin)
+				}
+			}
+		}
+		var wg sync.WaitGroup
+		defer EndTorpedoTest()
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:6404]: make this as first statement

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

**Comment by skonda-px** [comment]
Aetos DashBoard Link: http://aetos.pwx.purestorage.com/resultSet/testSetID/85961

**Review by snigdha-px — APPROVED** [review]
LGTM

File: tests/backup/backup_test.go [hunk 10]
```diff
@@ -6200,6 +6200,243 @@ var _ = Describe("{DeleteAllBackupObjects}", func() {
...
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			err = CreateSourceAndDestClusters(orgID, "", "", ctx)
+			log.FailOnError(err, "Source and Destination cluster creation failed")
+			clusterStatus, clusterUid = Inst().Backup.RegisterBackupCluster(orgID, SourceClusterName, "")
+			dash.VerifyFatal(clusterStatus, api.ClusterInfo_StatusInfo_Online, fmt.Sprintf("Verifying backup cluster %s status", SourceClusterName))
+		})
+		Step("Taking backup of applications", func() {
+			log.InfoD("Taking backup of applications")
+			backupName = fmt.Sprintf("%s-%v", BackupNamePrefix, time.Now().Unix())
+			//backupNames = append(backupNames, backupName)
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:6282]: remove the comment if not requierd

File: tests/backup/backup_test.go [hunk 11]
```diff
@@ -6200,6 +6200,243 @@ var _ = Describe("{DeleteAllBackupObjects}", func() {
...
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			CreateBackup(backupName, SourceClusterName, BackupLocationName, backupLocationUID, []string{bkpNamespaces[0]},
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:6285]: handle error here

File: tests/backup/backup_test.go [hunk 12]
```diff
@@ -6200,6 +6200,243 @@ var _ = Describe("{DeleteAllBackupObjects}", func() {
...
+							}
+							resp, err := Inst().Backup.InspectRestore(ctxNonAdmin, restoreInspectRequest)
+							if err != nil {
+								return "", true, fmt.Errorf("Restore Inspect failed with error %s", err)
+							}
+							restoreResponseStatus := resp.GetRestore().GetStatus()
+							actual := restoreResponseStatus.GetStatus()
+							expected := api.RestoreInfo_StatusInfo_Success
+							if actual != expected {
+								log.InfoD("Status of %s - [%s]", restore, restoreResponseStatus.GetStatus())
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:6370]: use infof here. as it is in retry logic

File: tests/backup/backup_test.go [hunk 13]
```diff
@@ -6200,6 +6200,243 @@ var _ = Describe("{DeleteAllBackupObjects}", func() {
...
+				defer wg.Done()
+				err := backup.DeleteUser(userName)
+				log.FailOnError(err, "Error deleting user %v", userName)
+			}(userName)
+		}
+		wg.Wait()
+		// log.Infof("Cleaning up backup location - %s", BackupLocationName)
+		// DeleteBackupLocation(BackupLocationName, backupLocationUID, orgID)
+
+		// log.Infof("Cleaning cloud credential")
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:6433]: remove the commented code if not required

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

**Comment by skonda-px** [comment]
http://aetos.pwx.purestorage.com/resultSet/testSetID/86003

**Review by lsrinivas-pure — APPROVED** [review]
LGTM, remove commented lines before merging

**Pull request merged**